From a509ec4f97129528e737315603514ef0c32e1265 Mon Sep 17 00:00:00 2001
From: Haihua Hu <jared.hu@nxp.com>
Date: Thu, 17 Dec 2020 12:11:45 +0800
Subject: [PATCH 56/91] gst/playback/: support runtime change connection-speed
 of adaptivedemux

update connection-speed at runtime in playbin, uridecodebin and decodebin
also do the same thing in playbin3/uridecodebin3/urisourcebin

Upstream-Status: Pending
https://gitlab.freedesktop.org/gstreamer/gst-plugins-base/-/merge_requests/981

Conflict:
	gst/playback/gstplaybin3.c
	gst/playback/gsturidecodebin3.c
(cherry picked from commit 7d1879c796ca007d2d2757f3b0060e0dc95bc302)
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 gst/playback/gstplaybin2.c      |  2 +-
 gst/playback/gsturidecodebin.c  |  2 --
 gst/playback/gsturidecodebin3.c | 26 +++++++++++++++++++++++++-
 gst/playback/gsturisourcebin.c  | 28 ++++++++++++----------------
 4 files changed, 38 insertions(+), 20 deletions(-)

diff --git a/gst/playback/gstplaybin2.c b/gst/playback/gstplaybin2.c
index 7cef7d5..575892c 100644
--- a/gst/playback/gstplaybin2.c
+++ b/gst/playback/gstplaybin2.c
@@ -2473,6 +2473,7 @@ gst_play_bin_set_property (GObject * object, guint prop_id,
       GST_PLAY_BIN_LOCK (playbin);
       playbin->connection_speed = g_value_get_uint64 (value) * 1000;
       connection_speed = playbin->connection_speed;
+      GST_PLAY_BIN_UNLOCK (playbin);
       if (playbin->curr_group) {
         GST_SOURCE_GROUP_LOCK (playbin->curr_group);
         if (playbin->curr_group->uridecodebin) {
@@ -2489,7 +2490,6 @@ gst_play_bin_set_property (GObject * object, guint prop_id,
         }
         GST_SOURCE_GROUP_UNLOCK (playbin->next_group);
       }
-      GST_PLAY_BIN_UNLOCK (playbin);
       break;
     }
     case PROP_BUFFER_SIZE:
diff --git a/gst/playback/gsturidecodebin.c b/gst/playback/gsturidecodebin.c
index 86e20a9..2b358bd 100644
--- a/gst/playback/gsturidecodebin.c
+++ b/gst/playback/gsturidecodebin.c
@@ -828,13 +828,11 @@ gst_uri_decode_bin_set_connection_speed (GstURIDecodeBin * dec)
   GST_OBJECT_UNLOCK (dec);
 
   /* set the property on all decodebins now */
-  GST_URI_DECODE_BIN_LOCK (dec);
   for (walk = dec->decodebins; walk; walk = g_slist_next (walk)) {
     GObject *decodebin = G_OBJECT (walk->data);
 
     g_object_set (decodebin, "connection-speed", connection_speed / 1000, NULL);
   }
-  GST_URI_DECODE_BIN_UNLOCK (dec);
 }
 
 static void
diff --git a/gst/playback/gsturidecodebin3.c b/gst/playback/gsturidecodebin3.c
index 9e6ace9..8bb80e0 100644
--- a/gst/playback/gsturidecodebin3.c
+++ b/gst/playback/gsturidecodebin3.c
@@ -257,6 +257,8 @@ struct _GstURIDecodeBin3
   gint shutdown;
 
   GList *output_pads;           /* List of OutputPad */
+
+  GList *source_handlers;       /* List of SourceHandler */
 };
 
 static GstStateChangeReturn activate_play_item (GstPlayItem * item);
@@ -1525,6 +1527,9 @@ new_source_handler (GstURIDecodeBin3 * uridecodebin, GstPlayItem * item,
       g_signal_connect (handler->urisourcebin, "about-to-finish",
       (GCallback) src_about_to_finish_cb, handler);
 
+  uridecodebin->source_handlers =
+      g_list_append (uridecodebin->source_handlers, handler);
+
   handler->expected_pads = 1;
 
   return handler;
@@ -1557,6 +1562,20 @@ source_handler_set_eos (GstSourceHandler * handler)
   }
 }
 
+static void
+gst_uri_decode_bin3_set_connection_speed (GstURIDecodeBin3 * dec)
+{
+  GList *walk;
+
+  /* set the property on all decodebins now */
+  for (walk = dec->source_handlers; walk; walk = g_list_next (walk)) {
+    GstSourceHandler *handler = (GstSourceHandler *) (walk->data);
+
+    g_object_set (handler->urisourcebin, "connection-speed",
+        dec->connection_speed / 1000, NULL);
+  }
+}
+
 static void
 gst_uri_decode_bin3_set_property (GObject * object, guint prop_id,
     const GValue * value, GParamSpec * pspec)
@@ -1573,6 +1592,7 @@ gst_uri_decode_bin3_set_property (GObject * object, guint prop_id,
     case PROP_CONNECTION_SPEED:
       GST_OBJECT_LOCK (dec);
       dec->connection_speed = g_value_get_uint64 (value) * 1000;
+      gst_uri_decode_bin3_set_connection_speed (dec);
       GST_OBJECT_UNLOCK (dec);
       break;
     case PROP_BUFFER_SIZE:
@@ -1732,7 +1752,11 @@ free_source_handler (GstURIDecodeBin3 * uridecodebin,
   }
   if (handler->pending_buffering_msg)
     gst_message_unref (handler->pending_buffering_msg);
-  g_free (handler);
+
+  uridecodebin->source_handlers =
+      g_list_remove (uridecodebin->source_handlers, handler);
+  g_slice_free (GstSourceHandler, handler);
+
 }
 
 static GstSourceItem *
diff --git a/gst/playback/gsturisourcebin.c b/gst/playback/gsturisourcebin.c
index cafbc69..7004f65 100644
--- a/gst/playback/gsturisourcebin.c
+++ b/gst/playback/gsturisourcebin.c
@@ -580,29 +580,25 @@ static void
 gst_uri_source_bin_update_connection_speed (GstURISourceBin * urisrc)
 {
   guint64 speed = 0;
-  GList *iter;
 
-  if (!urisrc->is_adaptive) {
-    return;
-  }
+  GParamSpec *pspec = NULL;
+  GList *tmp;
 
   GST_OBJECT_LOCK (urisrc);
   speed = urisrc->connection_speed / 1000;
   GST_OBJECT_UNLOCK (urisrc);
 
-  GST_URI_SOURCE_BIN_LOCK (urisrc);
-  for (iter = urisrc->src_infos; iter; iter = iter->next) {
-    ChildSrcPadInfo *info = iter->data;
-    GParamSpec *pspec = NULL;
-    if (!info->demuxer)
-      continue;
-
-    pspec = g_object_class_find_property (G_OBJECT_GET_CLASS (info->demuxer),
-        "connection-speed");
-    if (pspec != NULL)
-      g_object_set (info->demuxer, "connection-speed", speed, NULL);
+  if (urisrc->is_adaptive) {
+    for (tmp = urisrc->src_infos; tmp; tmp = tmp->next) {
+      ChildSrcPadInfo *info = tmp->data;
+      if (info->demuxer) {
+        pspec = g_object_class_find_property (G_OBJECT_GET_CLASS (info->demuxer),
+            "connection-speed");
+        if (pspec != NULL)
+          g_object_set (info->demuxer, "connection-speed", speed, NULL);
+      }
+    }
   }
-  GST_URI_SOURCE_BIN_UNLOCK (urisrc);
 }
 
 static void
-- 
2.43.0

